src/main/java/com/google/devtools/build/lib/rules/proto/ProtoLangToolchain.java

```diff
@@ -45,7 +45,14 @@ public ConfiguredTarget create(RuleContext ruleContext)
                 ruleContext.attributes().get("command_line", Type.STRING),
                 ruleContext.getPrerequisite("plugin", FilesToRunProvider.class),
                 ruleContext.getPrerequisite("runtime"),
-                blacklistedProtos.build()))
+                // We intentionally flatten the NestedSet here.
+                //
+                // `providedProtoSources` are read during analysis, so flattening the set here once
+                // avoid flattening it in every `<lang>_proto_aspect` applied to `proto_library`
+                // targets. While this has the potential to use more memory than using a NestedSet,
+                // there are only a few `proto_lang_toolchain` targets in every build, so the impact
+                // on memory consumption should be neglectable.
+                providedProtoSources.build().toList()))
```

Comment:
Why do you need to flatten the NestedSet? If I understand this correctly, so that you can use overloaded ProtoLangToolchainProvider.create? I'd prefer if you remove the old implementation of create and just use NestedSet. I checked that there are no dependencies internally. If you're worried about other forks of Bazel, perhaps you could change the create method name?

Reply:
Friendly ping.

Reply:
No, this not for overloading `ProtoLangToolchainProvider#create()`. Today, every `<lang>_proto_library` has to flatten the set to figure out which of its `srcs` is in `blacklistedProtos()`, which does have performance overhead. We can avoid that by flattening it here once and propagate the excluded protos as list instead. We trade some (finite) memory for performance.